### PR TITLE
Fix broken column after cloning product

### LINKED ISSUE #12636: [BUU] Columns broken after cloning
## Description
On the new Bulk Edit Products screen (`/admin/products` with `admin_style_v3`),
And I have hidden some columns using the "Columns" dropdown,

When I select the product action Clone,
A new product appears in the table, but
**An extra grey column appears to the side of the table (see picture)**

## Expected Behavior
The column sizes should remain as they were.

## Actual Behaviour
as above

## Steps to Reproduce
as above
## Animated Gif/Screenshot
Screenshot 2024-07-02 at 7 47 50

## Workaround
Select all columns for display, or reload the page.

## Severity
bug-s4: it's annoying, but you can use it

https://github.com/openfoodfoundation/openfoodnetwork/wiki/Bug-severity

## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by kernal053**
#### What? Why?

- Closes #12636 

#### What should we test?

- On Bulk Edit Products screen(`/admin/products`), I have checked some hidden columns in "Columns" dropdown section. After cloning a product, the resulting new product appears with updated colspan cell.
- I have used MutationObserver as per discussion to check for new added product row and adjusted colspan accordingly.

Note: test with 100 per page to see if any noticeable difference in performance.

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes

Fix broken column after cloning product

### CODE REVIEW

**Review by rioug — COMMENTED** [review]
It looks good, thanks @kernal053 . I fee like we should some sort of test for this, @dacook do you think we can cover this in system test ?

File: app/webpacker/controllers/column_preferences_controller.js [hunk 1]
```diff
@@ -30,14 +32,39 @@ export default class ColumnPreferencesController extends Controller {
     this.table.classList.toggle(`hide-${name}`, !element.checked);
 
     // Reset cell colspans
-    const hiddenColCount = this.checkboxes.filter((checkbox)=> !checkbox.checked).length;
```
Comment (dacook) [inline app/webpacker/controllers/column_preferences_controller.js:33]: Before, this count was executed only once before looping through all colspan cells. Now it's running every time, which is less efficient. 
But.. I think it's not too expensive and is ok. I like how it's now shared with the observer.

**Review by dacook — APPROVED** [review]
Great, thank you. It was good to see a use of MutationObserver and understand how it works.

I note that it's slightly less efficient now, so will add a note to test using 100 per page in case it has a noticeable effect.

**Comment by dacook** [comment]
> do you think we can cover this in system test ?

I think it wouldn't be worth testing as a system test, because we'd have to be inspecting HTML to test it.

I would probably have added a Stimulus JS test for this. But as it's @kernal053's first contribution, I won't require it. Something to consider for next time (Here's another JS test we have for reference: https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/javascripts/stimulus/checked_controller_test.js)

**Review by rioug — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Hey @kernal053 

Thanks for addressing this issue!

This looks great now. I've cloned a product after importing 100 products, to probe for any performance influence, but could not spot anything significant.

Before this PR
image

After this PR
image

Looks good!
Merging.

**Pull request merged**